cfgmgr/coppmgr.cpp

```diff
@@ -0,0 +1,859 @@
+#include <fstream>
+#include "logger.h"
+#include "dbconnector.h"
+#include "producerstatetable.h"
+#include "tokenize.h"
+#include "ipprefix.h"
+#include "coppmgr.h"
+#include "exec.h"
+#include "shellcmd.h"
+#include "warm_restart.h"
+#include "json.hpp"
+
+using json = nlohmann::json;
+
+using namespace std;
+using namespace swss;
+
+static set<string> g_copp_init_set;
+static map<string, string> trap_feature_map = {
+    {COPP_TRAP_TYPE_SAMPLEPACKET, "sflow"}
+};
+
+void CoppMgr::parseInitFile(void)
+{
+    std::ifstream ifs(COPP_INIT_FILE);
+    if (ifs.fail())
+    {
+        return; 
+    }
+    json j = json::parse(ifs);
+    for(auto tbl = j.begin(); tbl != j.end(); tbl++)
+    {
+        string table_name = tbl.key();
+        json keys = tbl.value();
+        for (auto k = keys.begin(); k != keys.end(); k++)
+        {
+            string table_key = k.key();
+            json fvp = k.value();
+            vector<FieldValueTuple> fvs;
+
+            for (auto f = fvp.begin(); f != fvp.end(); f++)
+            {
+                FieldValueTuple fv(f.key(), f.value().get<std::string>());
+                fvs.push_back(fv);
+            }
+            if (table_name == CFG_COPP_TRAP_TABLE_NAME)
+            {
+                m_coppTrapInitCfg[table_key] = fvs;
+            }
+            else if (table_name == CFG_COPP_GROUP_TABLE_NAME)
+            {
+                m_coppGroupInitCfg[table_key] = fvs;
+            }
+        }
+    }
+}
+
+/* Check if the trap group has traps that can be installed only when
+ * feature is enabled
+ */
+bool CoppMgr::checkIfTrapGroupFeaturePending(string trap_group_name)
+{
+    std::vector<FieldValueTuple> fvs;
+
+    if (m_coppGroupRestrictedMap.find(trap_group_name) == m_coppGroupRestrictedMap.end())
+    {
+        return false;
+    } 
+
+    for (auto it: m_coppTrapIdTrapGroupMap)
+    {
+        if (it.second == trap_group_name)
+        {
+            /* At least one trap should be enabled to install the 
+             * trap group with restricted attributes
+             */ 
+            if (!isTrapDisabled(it.first))
+            {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+void CoppMgr::setFeatureTrapIdsStatus(string feature, bool enable)
+{
+    for (auto it: trap_feature_map)
+    {
+        if (it.second ==  feature)
+        {
+            if (enable)
+            {
+                if (m_coppTrapDisabledMap.find(it.first) != m_coppTrapDisabledMap.end())
+                {
+                    if (m_coppTrapIdTrapGroupMap.find(it.first) != m_coppTrapIdTrapGroupMap.end())
+                    {
+                        string trap_group = m_coppTrapIdTrapGroupMap[it.first];
+
+                        /* Enable only when the restricted group is already pending */
+                        if ((m_coppGroupRestrictedMap.find(trap_group) 
```

Comment:
This is already checked in the `checkIfTrapGroupFeaturePending `function. Please remove

Reply:
Done